cep-recipe-jinja.md

```diff
@@ -0,0 +1,320 @@
+
+## Jinja in the new recipe format
+
+The new recipe format uses a subset of Jinja. Specifically, only "variable" expressions are allowed (no blocks such as `set`, `for` loops, `if`/`else` blocks, ...).
+
+A Jinja expression in the new recipe format looks like the following:
+
+`${{ version }}`
+
+Or if a function is involved:
+
+`${{ compiler('c') }}`
+
+Jinja expressions are also used in `if` statements and in the `skip` field of a recipe. In both instances, the `${{ ... }}` syntax is omitted.
+
+E.g.:
+
+```yaml
+build:
+  skip:
+    - osx  # This is a Jinja expression!
+
+requirements:
+  build:
+    - if: win and cuda  # This is a Jinja expression!
+      then:
+        - cudatoolkit
+```
+
+## Variables in the recipe
+
+The variables that are available in the Jinja context in the recipe come from two sources: the "variant configuration" file or the "context" section of the recipe.
+
+### The context section
+
+The context is a dictionary at the top-level of a recipe that maps keys to scalar values. The keys can be used by accessing them as variables in the Jinja expressions. For example:
+
+```yaml
+context:
+  version: "1.0.5"
+
+package:
+  version: ${{ version }}
+```
+
+Context evaluation happens from top-to-bottom. That means a later value can reference an earlier one like so:
+
+```yaml
+context:
+  version: "1.0.5"
+  name_and_version: "pkg_${{ version }}"
+```
+
+### Variables from variant configuration
+
+Any variable specified in the variant configuration file can be used in a recipe by using it in a Jinja expression. For example, with a variant config like:
+
+```yaml
+cuda:
+  - yes
+  - no
+```
+
+This value can be used as follows, for example in an inline `if` expression:
+
+```yaml
+requirements:
+  host:
+    - ${{ "cudatoolkit" if cuda == "yes" }}
+```
+
+## Available Jinja functions
+
+## The compiler function
+
+The compiler function is used to stick together a compiler from `{lang}_compiler` and `{lang}_compiler_version`
+
+The function looks as follows:
+
+```yaml
+${{ compiler('c') }}
+```
+
+This would pull in the c_compiler and c_compiler_version from the variant config. The compiler function suffixes `{lang}_compiler` with the `target_platform` to render to something such as:
+
+```
+gcc_linux-64 8.9
+clang_osx-arm64 12
+msvc_win-64 19.29
+```
+
+The function `${{ compiler("foo") }}` thus evaluates to `{foo_compiler}_{target_platform} {foo_compiler_version}`.
+
+The variant config could look something like:
+
+```yaml
+c_compiler:
+  - gcc
+c_compiler_version:
+  - "8.9"
+cxx_compiler:
+  - clang
+cxx_compiler_version:
+  - "12"
+```
+
+> [!NOTE]
+> Default values for the `compiler` function
+>
+> If not further specified, the following values are used as default values:4
+>
+> ```yaml
+> linux:
+>  c: gcc
+>  cxx: gxx
+>  fortran: gfortran
+>  rust: rust
+> osx:
+>   c: clang
+>   cxx: clangxx
+>   fortran: gfortran
+>   rust: rust
+> win:
+>   c: vs2017
+>   cxx: vs2017
```

Comment:
It might be nicer to use unversioned `vs` here? This is a relatively recent [addition](https://github.com/conda-forge/vc-feedstock/blob/ef25bb03d65e5c7adc024c07cbb2ccbe46b562bd/recipe/meta.yaml#L223), but that way the version would be independent of the default name, like on unix.

Reply:
We are inheriting the standards from `conda-build` here. I am happy to change them, and using an non-versioned `vs` would probably be better (altough also potentially annoying when it moves since you have to re-install vs on the build machine).

Reply:
And lastly, not sure what Anaconda has available on the defaults channel ...

Reply:
> altough also potentially annoying when it moves since you have to re-install vs on the build machine

Generally, our vs packages in conda-forge will pick up the compilers that you have - we have a fallback to use the local version (& toolchain target) if our default activation fails. Also, I'm not saying to abolish the `2017 / 2019 / 2022` pins, just move them to `c{,xx}_compiler_version`, like for unix.

Reply:
Nice! Didn't know that.

Reply:
cc @jezdez @chenghlee what is the situation on Anaconda's channels here? Should we make `vs` the default?

Reply:
I would avoid defaults. I think default values in tools are counter intuitive and should be defined on a per ecosystem basis and not by tools. In the sense that conda-forge should define their default and Anaconda should define their defaults, same for all the other ecosystems.

Reply:
To clarify my thoughts: Defaults like this are (or can) help less experienced package builders, but it can also be very disturbing and annoying for more experienced package builders. It can also cause troubles for ecosystems that might want to opt-out of these defaults (for example because they want to be very explicit about what they support and how things should work).

I sympathize to the cause of making it easier to use the build tools, but there are things which are very ecosystem specific. I believe compilers is one of these things.

Reply:
The only way around this is that we make an endpoint in the repodata or so where we can easily download the `conda_build_config.yaml` for a specified channel to get these defaults per-channel. 

If we don't have any easy way it means that each user needs _at least_ two files per recipe which is already a bit annoying, no?

Reply:
So what should we do with default values? I continue to think that default values for compilers and compiler versions should not be part of the CEP.